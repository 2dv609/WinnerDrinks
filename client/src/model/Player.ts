/**
 * A player class for keeping track of the player names,
 * their score and whether they are currently active in the game or not.
 * @author Delfi Sehidic
 */

class Player {
    private _name: string = ''; // The name of the player as registered in the input
    isActive: boolean; // Is the player currently playing or paused for e.g. a bathroom visit?
    score: number;

    /**
     * Constructs a new player object.
     * @param name Must be at least 3 characters long and not exceed 10.
     */
    constructor(name: string) {
      this.name = name;
      this.isActive = true;
      this.score = 0;
    }
    /**
     * Overwrites the old score of the player and sets it to newScore.
     * @param newScore
     */
    setScore(newScore: number) {
      this.score = newScore;
    }
    /**
     * Updates the score. For example, if the score is 10 and points is 1, the score will now be 11.
     *
     * @param points Added to the existing score. Can be negative.
     */
    addScore(points: number) {
      this.score += points;
    }
    /**
     * Returns a string representation of the player.
     * @param debug If set to true, it also returns the score and activity of the player.
     * @returns Player name or full player stats.
     */
    toString(debug?:boolean): string {
      if (debug) {
        return `Name: ${this._name} Score: ${this.score} isActive: ${this.isActive}`;
      }
      return this._name;
    }

    /**
     * Setter for property _name.
     */
    set name(name: string) {
      if (name.length < 1 || name.length > 10) {
        throw Error('Name must be at least 1 characters and not exceed 10.');
      }
      if (name.split(' ').length > 1) {
        throw Error('Name can not contain any spaces.');
      }
      this._name = name;
    }

    /**
     * Getter for property _name.
     */
    get name() {
      return this._name
    }
}

export default Player;
