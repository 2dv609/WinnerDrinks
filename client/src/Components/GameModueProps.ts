import Player from '../model/Player';

interface GameModuleProps {
    players: Player[],
    getPlayers: (amount: number, players: Player[]) => Player[]; // Let the component decide how many players it needs. 
    addScore: (p: Player, score: number) => void; // Adds a score to a player
    makeWinnerAlert: (p: any) => void; // makes an Alert that corresponds to if there are one, multiple, or no winners
    chooseRandomNewGame: () => void; // chooses a random new game
}

export type TextGameModuleProps = {
    gameService: GameModuleProps,
    gameEvent: IBackToBack | IParty | ITrivia
}

export type AnimationGameModuleProps = {
gameService: GameModuleProps
}


//export default GameProps;