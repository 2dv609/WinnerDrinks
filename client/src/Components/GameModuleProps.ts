import Player from '../model/Player';

interface GameModuleProps {
    players: Player[]
    getPlayers: (amount: number, players: Player[]) => Player[]; // Let the component decide how many players it needs. 
    addScore: (p: Player) => void; // Adds a score to a player
    makeWinnerAlert: (p: Player | Player[] | null, message?: string) => void; // makes an Alert that corresponds to if there are one, multiple, or no winners
    chooseRandomNewGame: () => void; // chooses a random new game
    removeGameEvent: (gameEventId: string, gameEvents: GameEventAPI) => void
}

export type TextGameModuleProps = {
    gameService: GameModuleProps,
    gameEvent: any,
    currentPlayers: Player[]
}

export type AnimationGameModuleProps = {
    gameService: GameModuleProps,
    currentPlayers: Player[]
}
