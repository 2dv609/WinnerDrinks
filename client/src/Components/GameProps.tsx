import Player from '../Player';

interface GameProps {
    getPlayers: (amount: number) => Player[]; // Let the component decide how many players it needs. 
    addScore: (p: Player, score: number) => void; // Adds a score to a player
    makeWinnerAlert: (p: any, message?:string) => void; // makes an Alert that corresponds to if there are one, multiple, or no winners
    chooseRandomNewGame: () => void; // chooses a random new game
    updateScoreboard: () => void;
}

export default GameProps;