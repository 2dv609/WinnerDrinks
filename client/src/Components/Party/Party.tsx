import React from 'react';
import { TextGameModuleProps } from '../GameModuleProps';


const Party: React.FC<TextGameModuleProps> = ({ gameService, gameEvent, currentPlayers }) => {

    function addPlayerToGameEvent(): string {
        return gameEvent.question.replace(/{p0}/g, `${currentPlayers[0]}`).replace(/{p1}/g, `${currentPlayers[1]}`)
    }

    return (
        <div className="card block">
            <div className="card-content">
                <div className="content" data-testid="game-event">{addPlayerToGameEvent()}</div>
                <h3>Who won?</h3>
                <div className="columns">
                    <div className="column">
                        <button className="button" data-testid="button-1" onClick={() => {
                            gameService.addScore(currentPlayers[1])
                            gameService.makeWinnerAlert(currentPlayers[1])
                            gameService.chooseRandomNewGame()
                        }}>{currentPlayers[1].toString()}</button>
                    </div>
                    <div className="column">
                        <button className="button" data-testid="button-0" onClick={() => {
                            gameService.addScore(currentPlayers[0])
                            gameService.makeWinnerAlert(currentPlayers[0])
                            gameService.chooseRandomNewGame()
                        }}>{currentPlayers[0].toString()}</button>
                    </div>
                </div>




            </div>
        </div>

    )
};

export default Party;