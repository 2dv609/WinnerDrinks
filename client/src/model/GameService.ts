import Player from './Player'
import { TextGameModuleProps, AnimationGameModuleProps } from '../Components/GameModuleProps'

export default class GameService {

    public addScore = (p: Player, score: number) => {
        p.addScore(score)
    }

    public getNewGameIndex = (currentGameIndex: number, gameModules: (React.FC<TextGameModuleProps> | React.FC<AnimationGameModuleProps>)[], activeGames: any[]): number => {
        let newIndex = currentGameIndex
        while (newIndex === currentGameIndex || !activeGames[newIndex].active) { // Don't allow the same game twice in a row. 
          console.log('while')
          newIndex = Math.floor(Math.random() * gameModules.length)
        }
        return newIndex
    }
    /**
     * 
     * @param nrOfPlayers The amount of players you need for the component
     * @param players An array of players that the function will fetch from. 
     * @returns A sublist from your array that contains the requested amount of active players. 
     */
    public getPlayers = (nrOfPlayers: number, players: Player[]): Player[] => {
        const activePlayers: Player[] = [];
        
        for (let i = 0; i < players.length; i++) {
            if (players[i].isActive) {
                activePlayers.push(players[i])
            }
            
        }

        this.shuffle(activePlayers);
        
        return activePlayers.slice(0, nrOfPlayers);
    }

    private shuffle(array: Player[]) {
        var m = array.length, t, i;
        while (m) {
            i = Math.floor(Math.random() * m--);
            t = array[m];
            array[m] = array[i];
            array[i] = t;
        }
        return array;
    }

    public getRandomGameEvent = (gameEventAPI: GameEventAPI): IBackToBack | IParty | ITrivia => {
        return gameEventAPI.questions[Math.floor(Math.random() * gameEventAPI.questions.length)]
    }

    public getNumActivePlayers = (players: Player[]): number => {
        let numOfActivePlayers = 0

        players.forEach(player => {
            if(player.isActive) numOfActivePlayers++  
        })

        return numOfActivePlayers
      }
}