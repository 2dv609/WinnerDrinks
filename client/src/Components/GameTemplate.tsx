import { useState, useEffect } from 'react';
import GameProps from './GameProps';
import './../App.css';

const NUM_OF_PLAYERS = 2; // How many players do you need? 

function MyGame(props: any) {
    const gp: GameProps = props.gp; // Contains the methods done and getPlayers. 
    const [players, setPlayers] = useState(gp.getPlayers(NUM_OF_PLAYERS)); // players is an array 

    useEffect(() => { 
        return () => { // Return a function for code cleanup. This will set new players 

            setPlayers(gp.getPlayers(NUM_OF_PLAYERS));
        }
    }, [gp])

    // Do some logic here. Use the array players[someNumber] and when you're done,
    // call the function gp.done() and send in the winning player and some score. 

    return (
        <div>
            <h3>Selected players</h3>
            <ul>
                {/* Just make a list of the players */}
            {players.map(pl => (<li>{pl.toString()}</li>))} 
            </ul>
            {/* This buttons makes the player first in the list win, and adds 0 to their score. */}
            <button onClick={() => 
                doneReplacementFunction()
                } >Click here to win.</button>
        </div>
    )

    function doneReplacementFunction() {
        gp.addScore(players[0], 0)
        gp.makeWinnerAlert(players[0])
        gp.chooseRandomNewGame()
    }
};

export default MyGame;