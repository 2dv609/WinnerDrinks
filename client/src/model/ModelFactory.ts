import GameModuleService from './GameModuleService'
import IGameModuleService from './IGameModuleService'
import GameService from './GameService'
import { IGameModuleSetting } from '../Components/Menu/Navbar'

/**
 * Factory method that return a promise of an instance 
 * of a type IGameModuleSerice with loaded game events or
 * undefined if GameModules coulde not be loaded.
 * 
 * @returns {Promis<IUtilservice | undefined>}
 */
export async function getGameModuleService(): Promise<IGameModuleService | undefined> {

    const gameModuleSerivce: IGameModuleService = new GameModuleService()
    const gameEventIsLoaded: boolean = await gameModuleSerivce.loadGameEvents()

    if (!gameEventIsLoaded) {
        return
    }

    return gameModuleSerivce
}

/**
 * Factory method that return  an instance 
 * of a type IGameService.
 * 
 * @returns {GameService}
 */
 export function getGameService(): GameService {
    return new GameService()
}

export function getGameModuleSettings(): IGameModuleSetting[] {
    
    return [
        { name: 'Wheel', enable: true, active: true, index: 0 },
        { name: 'Party', enable: true, active: true, index: 1 },
        { name: 'BackToBack', enable: false, active: false, index: 2 },
        { name: 'Trivia', enable: true, active: true, index: 3 },
    ]
}
