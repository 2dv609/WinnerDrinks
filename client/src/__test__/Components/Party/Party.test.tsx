import React from 'react';
import ReactDOM from 'react-dom';
import Party from '../../../Components/Party/Party'
import { render, cleanup, fireEvent } from '@testing-library/react' 
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'
import { getGameService } from '../../../model/ModelFactory'
import GameService from '../../../model/GameService'
import { GameModuleProps } from '../../../Components/GameModuleProps'
import Player from '../../../model/Player'

/* --------------------------------- */
/* Test cases for game module party. */
/* --------------------------------- */

describe('Party test', () => {

    /* --------------------------------- */
    /* ----- Test cases setup ---------- */
    /* --------------------------------- */

    const players: Player[] = []
    let gameService: GameService
    let gameModuleProps: GameModuleProps
    const makeWinnerAlertMock = (p: Player | Player[] | null, message?: string): void => {
        console.log('makeWinnerAlertMock')
    }
    const chooseRandomNewGameMock = (): void => {
        console.log('chooseRandomNewGameMock')
    }
    const removeGameEventMock = (gameEventId: string, gameEvents: GameEventAPI): void => {}
    let currentPlayers: Player[] = []
    const currentQuestionMock: IParty = {
        _id: 'mockID',
        question: 'mockQuestion {p0} and {p1}'
    } 

    beforeAll(() => {
        gameService = getGameService()
        players.push(new Player('Tom'), new Player('Carl'), new Player('Sandra'))
        currentPlayers = gameService.getPlayers(2, players)  
        gameModuleProps = {
        players: players,
        getPlayers: gameService.getPlayers,
        addScore: gameService.addScore,
        makeWinnerAlert: makeWinnerAlertMock,
        chooseRandomNewGame: chooseRandomNewGameMock,
        removeGameEvent: removeGameEventMock
      }
    })

    afterEach(cleanup)

    test('Game module props should use props gameService, gameEvent and currentPlayers', () => {
        const div = document.createElement('div')
        ReactDOM.render(<Party currentPlayers={currentPlayers} gameService={gameModuleProps} gameEvent={currentQuestionMock}></Party>, div)
    })

    test('T1.MP.UI.1: Game module party should display a party game event', () => {
        const { getByTestId } = render(<Party currentPlayers={currentPlayers} gameService={gameModuleProps} gameEvent={currentQuestionMock}></Party>)
        expect(getByTestId('game-event')).toHaveTextContent(addPlayerToGameEvent(currentQuestionMock.question, currentPlayers))
        expect(getByTestId('button-0')).toHaveTextContent(currentPlayers[0].name)
        expect(getByTestId('button-1')).toHaveTextContent(currentPlayers[1].name)
    })

    test('T1.MP.UI.2: Party game events should contain 2 active game participants names', () => {
        const { getByTestId } = render(<Party currentPlayers={currentPlayers} gameService={gameModuleProps} gameEvent={currentQuestionMock}></Party>)
        const gameEvent: string | null = getByTestId('game-event').textContent
        expect(isPlayerNamesIncludedInGameEvent(currentPlayers, gameEvent)).toBeTruthy()
    })

    test('T1.MP.UI.3: A user should be able to select a party game event winner, button 0', () => {
        const { getByTestId } = render(<Party currentPlayers={currentPlayers} gameService={gameModuleProps} gameEvent={currentQuestionMock}></Party>)
        
        const addScoreSpy = jest.spyOn(gameModuleProps, 'addScore')
        const makeWinnerAlertSpy = jest.spyOn(gameModuleProps, 'makeWinnerAlert')
        const chooseNewRandomGameSpy = jest.spyOn(gameModuleProps, 'chooseRandomNewGame')
        
        fireEvent.click(getByTestId('button-0'))
        expect(addScoreSpy).toBeCalled()
        expect(makeWinnerAlertSpy).toBeCalled()
        expect(chooseNewRandomGameSpy).toBeCalled()
    })

    test('T1.MP.UI.3: A user should be able to select a party game event winner, button 1', () => {
        const { getByTestId } = render(<Party currentPlayers={currentPlayers} gameService={gameModuleProps} gameEvent={currentQuestionMock}></Party>)
        
        const addScoreSpy = jest.spyOn(gameModuleProps, 'addScore')
        const makeWinnerAlertSpy = jest.spyOn(gameModuleProps, 'makeWinnerAlert')
        const chooseNewRandomGameSpy = jest.spyOn(gameModuleProps, 'chooseRandomNewGame')
        
        fireEvent.click(getByTestId('button-1'))
        expect(addScoreSpy).toBeCalled()
        expect(makeWinnerAlertSpy).toBeCalled()
        expect(chooseNewRandomGameSpy).toBeCalled()
    })

    test('Game module tree should match snapshot', () => {
        const tree = renderer.create(<Party currentPlayers={currentPlayers} gameService={gameModuleProps} gameEvent={currentQuestionMock}></Party>).toJSON()
        expect(tree).toMatchSnapshot()
    })
})

/* --------------------------------- */
/* ----- Test help functions ------- */
/* --------------------------------- */

function addPlayerToGameEvent(question: string, currentPlayers: Player[]): string {
    return question.replace('{p0}', currentPlayers[0].name).replace('{p1}', `${currentPlayers[1].name}`)
}

function isPlayerNamesIncludedInGameEvent(players: Player[], gameEvent: string | null): boolean {
    if (!gameEvent) {
        return false
    }
    return players.every((player: Player) => gameEvent.includes(player.name))
}
