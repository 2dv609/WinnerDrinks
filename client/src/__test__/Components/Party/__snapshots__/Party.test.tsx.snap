// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Party test Game module tree should match snapshot 1`] = `
<div
  className="card block"
>
  <div
    className="card-content"
  >
    <div
      className="content"
      data-testid="game-event"
    >
      mockQuestion Sandra and Carl
    </div>
    <h3>
      Who won?
    </h3>
    <div
      className="columns"
    >
      <div
        className="column"
      >
        <button
          className="button"
          data-testid="button-1"
          onClick={[Function]}
        >
          Carl
        </button>
      </div>
      <div
        className="column"
      >
        <button
          className="button"
          data-testid="button-0"
          onClick={[Function]}
        >
          Sandra
        </button>
      </div>
    </div>
  </div>
</div>
`;
